### ISSUE TITLE
error occur when I make my own vocabulary instead of marian vocabulary

### ISSUE THREAD

**Issue by wingsyuan** [marian-nmt/marian-dev#445]
Hi,
when I use the general vocabulary and use the command awk '{print $0": "FNR - 1}' to add num to the word per line,  there comes the problem:
</s>: 0
<unk>: 1
`: 2
`@@: 3
~: 4
~(: 5
~@@: 6
÷: 7
×: 8
××: 9
×@@: 10
<: 11
<<: 12

Error: Unhandled exception of type 'N4YAML15ParserExceptionE': yaml-cpp: error at line 8, column 1: end of map not found
[2019-05-04 14:54:04] Error: Aborted from void unhandledException() in /home/mt/marian-dev/src/common/logging.cpp:107

 How did the problem come up? thanks very much!!!

**Comment by frankseide** [comment]
Marian supports plain-text files. I think you can just feed it your own vocab file as-is. The advantage of plain-text over YAML is that YAML vocabs only work if your input data is UTF-8 encoded (and does not contain invalid UTF-8, which some of the typical benchmark data sets do, I believe).

Plain-text vocabs:


  *   one word entry per line, no escaping, no tabs/spaces/comments allowed. The internal word index is the line number (first line has index 0). CRLF line-ending is allowed, in case you are on Windows.
  *   the first three tokens are usually `<unk/>`, `<s>` and `</s>`
  *   pathname extension must not be one of the typical extensions for Yaml or Json. I don’t have the source code in front of me, but I think any other extension else is just fine. I would use .wl (for word list) or .txt

**Comment by wingsyuan** [comment]
Hi,
@frankseide  thanks for your help!

Can I ask you another question?
when I use the same training dataset  with only different vocabulary, the results show different too:

First result:
[2019-05-05 23:28:51] [valid] Ep. 1 : Up. 10500 : translation : 10.4 : new best
[2019-05-05 23:32:38] Ep. 1 : Up. 11000 : Sen. 4,270,605 : Cost 3.51933360 : Time 356.49s : 14431.62 words/s : L.r. 2.0625e-04
[2019-05-05 23:36:25] Ep. 1 : Up. 11500 : Sen. 4,470,838 : Cost 3.46137643 : Time 226.71s : 22832.60 words/s : L.r. 2.1563e-04
[2019-05-05 23:40:10] Ep. 1 : Up. 12000 : Sen. 4,659,743 : Cost 3.40844083 : Time 224.70s : 22674.27 words/s : L.r. 2.2500e-04
[2019-05-05 23:40:10] Saving model weights and runtime parameters to model_out/ens/model.npz.orig.npz
[2019-05-05 23:40:16] Saving model weights and runtime parameters to model_out/ens/model.npz
[2019-05-05 23:40:22] Saving Adam parameters to model_out/ens/model.npz.optimizer.npz
[2019-05-05 23:40:34] Saving model weights and runtime parameters to model_out/ens/model.npz.best-ce-mean-words.npz
[2019-05-05 23:40:37] [valid] Ep. 1 : Up. 12000 : ce-mean-words : 4.25314 : new best
[2019-05-05 23:40:38] Saving model weights and runtime parameters to model_out/ens/model.npz.best-perplexity.npz
[2019-05-05 23:40:42] [valid] Ep. 1 : Up. 12000 : perplexity : 70.3257 : new best
[2019-05-05 23:41:48] Saving model weights and runtime parameters to model_out/ens/model.npz.best-translation.npz
[2019-05-05 23:41:52] [valid] Ep. 1 : Up. 12000 : translation : 11.55 : new best
[2019-05-05 23:45:40] Ep. 1 : Up. 12500 : Sen. 4,859,206 : Cost 3.34525561 : Time 329.89s : 15743.96 words/s : L.r. 2.3438e-04
[2019-05-05 23:47:09] Seen 4934624 samples

the second:

[2019-05-06 11:04:01] [valid] Ep. 1 : Up. 10500 : ce-mean-words : 4.3597 : new best
[2019-05-06 11:04:02] Saving model weights and runtime parameters to model_out_lawvoc/ens/model.npz.best-perplexity.npz
[2019-05-06 11:04:05] [valid] Ep. 1 : Up. 10500 : perplexity : 78.2337 : new best
[2019-05-06 11:05:27] Saving model weights and runtime parameters to model_out_lawvoc/ens/model.npz.best-translation.npz
[2019-05-06 11:05:31] [valid] Ep. 1 : Up. 10500 : translation : 10.95 : new best
[2019-05-06 11:06:37] Seen 4934624 samples
[2019-05-06 11:06:37] Starting epoch 2
[2019-05-06 11:06:37] Training finished

Why the second result show  training update 10500 steps had seed 4934624 samples but the first result  "Seen 4934624 samples" when update 12000 steps?why are they different? 

Thanks a lot!!

**Comment by snukky** [comment]
4,934,624 is the number of processed training sentences after the 1st epoch, and there is no information how many updates exactly was in this epoch (we only know that it was more than 12,500 in the first training run and more than 10,500 in the other). With `--mini-batch-fit` batch sizes are dynamic, determined by sentence lengths. Different vocabularies mean different segmentation (I assume you use BPE) and usually sentence lengths.

**Comment by frankseide** [comment]
You can get more detailed logging with this option `--disp-label-counts`. This gives you the precise number of target-word labels used by a specific model iteration. The output looks like this:
```
Ep. 3 : Up. 370000 : Sen. 12,803,458 : Cost 3.12760186 * 116,836,225 @ 26,880
    after 8,646,852,040 : Time 2291.29s : 50991.50 words/s : L.r. 2.9409e-05
```
(I broken the line manually for better readability in this thread viewer). The string `after 8,646,852,040`  means that this model has seen 8.6 billion labels so far.

Note, however, that with `--mini-batch-fit`, this does not guarantee that if two trainings with different model dimensions have seen the same number of labels that they have seen the same set of training samples. If the batch size changes due to `--mini-batch-fit`, the batching itself also changes (since we try to fill batches with similar-length sentences). It is, however, likely that they did anyway, since is very unlikely to get identical label counts unless batching ended up being the same.

**Comment by emjotde** [comment]
Should this stay open? @frankseide

**Comment by emjotde** [comment]
Closing due to inactivity.
